plugins/modules/git_config.py

```diff
@@ -197,94 +217,127 @@ def main():
-        unset = 'unset'
-        params['value'] = None
-    else:
-        unset = None
-
-    if params['value']:
-        new_value = params['value']
-    else:
-        new_value = None
+    base_args = [git_path, "config", "--includes"]
 
-    args = [git_path, "config", "--includes"]
-    if params['list_all']:
-        args.append('-l')
     if scope == 'file':
-        args.append('-f')
-        args.append(params['file'])
+        base_args.append('-f')
+        base_args.append(params['file'])
     elif scope:
-        args.append("--" + scope)
+        base_args.append("--" + scope)
+
+    list_args = list(base_args)
+
+    if params['list_all']:
+        list_args.append('-l')
+
     if name:
-        args.append(name)
+        list_args.append("--get-all")
+        list_args.append(name)
 
-    if scope == 'local':
-        dir = params['repo']
-    elif params['list_all'] and params['repo']:
-        # Include local settings from a specific repo when listing all available settings
-        dir = params['repo']
-    else:
-        # Run from root directory to avoid accidentally picking up any local config settings
-        dir = "/"
+    (rc, out, err) = module.run_command(list_args, cwd=cwd, expand_user_and_vars=False)
 
-    (rc, out, err) = module.run_command(args, cwd=dir, expand_user_and_vars=False)
     if params['list_all'] and scope and rc == 128 and 'unable to read config file' in err:
         # This just means nothing has been set at the given scope
         module.exit_json(changed=False, msg='', config_values={})
     elif rc >= 2:
         # If the return code is 1, it just means the option hasn't been set yet, which is fine.
-        module.fail_json(rc=rc, msg=err, cmd=' '.join(args))
+        module.fail_json(rc=rc, msg=err, cmd=' '.join(list_args))
+
+    old_values = out.rstrip().splitlines()
 
     if params['list_all']:
-        values = out.rstrip().splitlines()
         config_values = {}
-        for value in values:
+        for value in old_values:
             k, v = value.split('=', 1)
-            config_values[k] = v
+            if k in config_values:
+                config_values[k].append(v)
+            else:
+                config_values[k] = [v]
+
+        # All values in the dict are lists. If a list only contains one element, replace the
+        # list by the element itself.
+        for k, v in config_values.items():
+            if len(v) == 1:
+                config_values[k] = str(v[0])
         module.exit_json(changed=False, msg='', config_values=config_values)
     elif not new_value and not unset:
-        module.exit_json(changed=False, msg='', config_value=out.rstrip())
+        module.exit_json(changed=False, msg='', config_value=old_values[0] if len(old_values) == 1 else old_values or '')
     elif unset and not out:
         module.exit_json(changed=False, msg='no setting to unset')
+    elif new_value in old_values:
+        module.exit_json(changed=False, msg="")
```

Comment:
Shouldn't this consider `add_mode`? If it is `replace-all`, I would expect that the other values are removed. (Or is this a breaking change for the module? In that case, maybe `add_mode` should be renamed to `value_mode` and have yet another option that allows to clear all other values.)

Reply:
Good point. Done